### PR TITLE
[NeMo-UX] Fix a serialization bug affecting fsdp

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

This PR recursively finds the leaves of the trainer config and calls `track_io` on the leaves. This handles the case of adding io functionality to plugins which are dictionaries. 

**Collection**: llm

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by blahBlahhhJ — APPROVED** [review]

File: nemo/lightning/pytorch/trainer.py [hunk 1]
```diff
@@ -10,18 +10,23 @@
 
 
 class Trainer(pl.Trainer, IOMixin):
+
+    def add_io(self, obj):
```
Comment (marcromeyn) [inline nemo/lightning/pytorch/trainer.py:14]: This function can be improved. You can use isinstance checks with the `Sequence` + `Mapping` types. I think it's easier to read if we start with that and move the top part into an else.
